File: samples/java/Bingo/Server/Api/src/main/java/systems/zlink/samples/bingo/server/api/handlers/ReportBingoResultHandler.java
package systems.zlink.samples.bingo.server.api.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.bingo.server.api.BingoPlayerRecordStore;
import systems.zlink.samples.bingo.server.configuration.SampleNames;
import systems.zlink.samples.bingo.shared.contracts.BingoMessages;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup(SampleNames.ApiChannel)
public final class ReportBingoResultHandler
        implements ZLinkRequestHandler<
                Messages.ReportBingoResultReq, Messages.ReportBingoResultRes> {
    private final BingoPlayerRecordStore records;

    public ReportBingoResultHandler(BingoPlayerRecordStore records) {
        this.records = records;
    }

    @Override
    public CompletionStage<Messages.ReportBingoResultRes> handle(
            Messages.ReportBingoResultReq request, ZLinkMessageContext context) {
        var record = records.report(request.getActorId(), request.getWon());
        return CompletableFuture.completedFuture(
                BingoMessages.reportBingoResultRes(
                        record.actorId(), record.wins(), record.losses()));
    }
}

File: samples/kotlin/Bingo/Server/Api/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/api/handlers/ReportBingoResultHandler.kt
package systems.zlink.samples.kotlin.bingo.server.api.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.samples.kotlin.bingo.server.api.BingoPlayerRecordStore
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.shared.contracts.ReportBingoResultReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.ReportBingoResultRes

@ZLinkHandlerGroup(SampleNames.ApiChannel)
class ReportBingoResultHandler(private val records: BingoPlayerRecordStore) :
    ZLinkSuspendingRequestHandler<ReportBingoResultReq, ReportBingoResultRes> {
    override suspend fun handle(
        request: ReportBingoResultReq,
        context: ZLinkMessageContext,
    ): ReportBingoResultRes {
        val record = records.report(request.actorId, request.won)
        return ReportBingoResultRes(record.actorId, record.wins, record.losses)
    }
}
